DATA RACE in the TestReleaseWhenRunningPool

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)

```
==================
WARNING: DATA RACE
Write at 0x00c000261228 by goroutine 20:
  runtime.racewrite()
      <autogenerated>:1 +0x24
  github.com/pingcap/tidb/resourcemanager/pool/spool.(*Pool).ReleaseAndWait()
      resourcemanager/pool/spool/spool.go:176 +0x48
  github.com/pingcap/tidb/resourcemanager/pool/spool.TestReleaseWhenRunningPool()
      resourcemanager/pool/spool/spool_test.go:64 +0x224
  testing.tRunner()
      GOROOT/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      GOROOT/src/testing/testing.go:1629 +0x47
Previous read at 0x00c000261228 by goroutine 21:
  runtime.raceread()
      <autogenerated>:1 +0x24
  github.com/pingcap/tidb/resourcemanager/pool/spool.(*Pool).run()
      resourcemanager/pool/spool/spool.go:109 +0x3c
  github.com/pingcap/tidb/resourcemanager/pool/spool.(*Pool).Run()
      resourcemanager/pool/spool/spool.go:90 +0xa6
  github.com/pingcap/tidb/resourcemanager/pool/spool.TestReleaseWhenRunningPool.func1()
      resourcemanager/pool/spool/spool_test.go:50 +0xaf
Goroutine 20 (running) created at:
  testing.(*T).Run()
      GOROOT/src/testing/testing.go:1629 +0x805
  testing.runTests.func1()
      GOROOT/src/testing/testing.go:2036 +0x8d
  testing.tRunner()
      GOROOT/src/testing/testing.go:1576 +0x216
  testing.runTests()
      GOROOT/src/testing/testing.go:2034 +0x87c
  testing.(*M).Run()
      GOROOT/src/testing/testing.go:1906 +0xb44
  go.uber.org/goleak.VerifyTestMain()
      external/org_uber_go_goleak/testmain.go:53 +0x70
  github.com/pingcap/tidb/resourcemanager/pool/spool.TestMain()
      resourcemanager/pool/spool/main_test.go:32 +0x435
  main.main()
      bazel-out/k8-fastbuild/bin/resourcemanager/pool/spool/spool_test_/testmain.go:110 +0x5e8
Goroutine 21 (running) created at:
  github.com/pingcap/tidb/resourcemanager/pool/spool.TestReleaseWhenRunningPool()
      resourcemanager/pool/spool/spool_test.go:47 +0x164
  testing.tRunner()
      GOROOT/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      GOROOT/src/testing/testing.go:1629 +0x47
==================
--- FAIL: TestReleaseWhenRunningPool (0.01s)
    testing.go:1446: race detected during execution of test
panic: sync: WaitGroup is reused before previous Wait has returned [recovered]
	panic: sync: WaitGroup is reused before previous Wait has returned
goroutine 35 [running]:
testing.tRunner.func1.2({0x18fe2e0, 0x1f6da60})
	GOROOT/src/testing/testing.go:1526 +0x372
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1529 +0x650
panic({0x18fe2e0, 0x1f6da60})
	GOROOT/src/runtime/panic.go:890 +0x263
sync.(*WaitGroup).Wait(0xc000261220)
	GOROOT/src/sync/waitgroup.go:118 +0x107
github.com/pingcap/tidb/resourcemanager/pool/spool.(*Pool).ReleaseAndWait(0xc000261220)
	resourcemanager/pool/spool/spool.go:176 +0x49
github.com/pingcap/tidb/resourcemanager/pool/spool.TestReleaseWhenRunningPool(0x0?)
	resourcemanager/pool/spool/spool_test.go:64 +0x225
testing.tRunner(0xc000307ba0, 0x1b7ec50)
	GOROOT/src/testing/testing.go:1576 +0x217
created by testing.(*T).Run
	GOROOT/src/testing/testing.go:1629 +0x806 
```
### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

--- a/resourcemanager/pool/spool/BUILD.bazel
+++ b/resourcemanager/pool/spool/BUILD.bazel
@@ -30,6 +30,7 @@ go_test(
         "spool_test.go",
     ],
     embed = [":spool"],
+    flaky = True,
     race = "on",
     shard_count = 2,
     deps = [

--- a/resourcemanager/pool/spool/BUILD.bazel
+++ b/resourcemanager/pool/spool/BUILD.bazel
@@ -30,6 +30,7 @@ go_test(
         "spool_test.go",
     ],
     embed = [":spool"],
+    race = "on",
     shard_count = 2,
     deps = [
         "//resourcemanager/pool",

--- a/resourcemanager/pool/spool/spool.go
+++ b/resourcemanager/pool/spool/spool.go
@@ -185,6 +185,7 @@ func (p *Pool) ReleaseAndWait() {
 	p.isStop.Store(true)
 	// wait for all the task in the pending to exit
 	for p.waiting.Load() > 0 {
+		_ = "" // it is empty statement to skip empty-block linter.
 	}
 	p.wg.Wait()
 	resourcemanager.InstanceResourceManager.Unregister(p.Name())

--- a/resourcemanager/pool/spool/spool.go
+++ b/resourcemanager/pool/spool/spool.go
@@ -145,6 +145,9 @@ func (p *Pool) RunWithConcurrency(fns chan func(), concurrency uint32) error {
 // checkAndAddRunning is to check if a task can run. If can, add the running number.
 func (p *Pool) checkAndAddRunning(concurrency uint32) (conc int32, run bool) {
 	for {
+		if p.isStop.Load() {
+			return 0, false
+		}
 		p.mu.Lock()
 		value, run := p.checkAndAddRunningInternal(int32(concurrency))
 		if run {
@@ -156,9 +159,6 @@ func (p *Pool) checkAndAddRunning(concurrency uint32) (conc int32, run bool) {
 			return 0, false
 		}
 		p.mu.Unlock()
-		if p.isStop.Load() {
-			return 0, false
-		}
 		time.Sleep(waitInterval)
 	}
 }

--- a/resourcemanager/pool/spool/spool.go
+++ b/resourcemanager/pool/spool/spool.go
@@ -40,6 +40,7 @@ type Pool struct {
 	options            *Options
 	capacity           int32
 	running            atomic.Int32
+	waiting            atomic.Int32
 	isStop             atomic.Bool
 	concurrencyMetrics prometheus.Gauge
 	taskManager        pooltask.TaskManager[any, any, any, any, pooltask.NilContext]
@@ -144,6 +145,8 @@ func (p *Pool) RunWithConcurrency(fns chan func(), concurrency uint32) error {
 
 // checkAndAddRunning is to check if a task can run. If can, add the running number.
 func (p *Pool) checkAndAddRunning(concurrency uint32) (conc int32, run bool) {
+	p.waiting.Add(1)
+	defer p.waiting.Add(-1)
 	for {
 		if p.isStop.Load() {
 			return 0, false
@@ -179,7 +182,9 @@ func (p *Pool) checkAndAddRunningInternal(concurrency int32) (conc int32, run bo
 func (p *Pool) ReleaseAndWait() {
 	p.isStop.Store(true)
 	// wait for all the task in the pending to exit
-	time.Sleep(2 * waitInterval)
+	for p.waiting.Load() > 0 {
+		time.Sleep(waitInterval)
+	}
 	p.wg.Wait()
 	resourcemanager.InstanceResourceManager.Unregister(p.Name())
 }

--- a/resourcemanager/pool/spool/spool.go
+++ b/resourcemanager/pool/spool/spool.go
@@ -31,6 +31,8 @@ import (
 	"go.uber.org/zap"
 )
 
+const waitInterval = 5 * time.Millisecond
+
 // Pool is a single goroutine pool. it can not reuse the goroutine.
 type Pool struct {
 	wg                 sync.WaitGroup
@@ -154,7 +156,10 @@ func (p *Pool) checkAndAddRunning(concurrency uint32) (conc int32, run bool) {
 			return 0, false
 		}
 		p.mu.Unlock()
-		time.Sleep(5 * time.Millisecond)
+		if p.isStop.Load() {
+			return 0, false
+		}
+		time.Sleep(waitInterval)
 	}
 }
 
@@ -173,6 +178,8 @@ func (p *Pool) checkAndAddRunningInternal(concurrency int32) (conc int32, run bo
 // ReleaseAndWait releases the pool and waits for all tasks to be completed.
 func (p *Pool) ReleaseAndWait() {
 	p.isStop.Store(true)
+	// wait for all the task in the pending to exit
+	time.Sleep(2 * waitInterval)
 	p.wg.Wait()
 	resourcemanager.InstanceResourceManager.Unregister(p.Name())
 }

--- a/resourcemanager/pool/spool/spool.go
+++ b/resourcemanager/pool/spool/spool.go
@@ -42,6 +42,7 @@ type Pool struct {
 	running            atomic.Int32
 	waiting            atomic.Int32
 	isStop             atomic.Bool
+	cond               sync.Cond
 	concurrencyMetrics prometheus.Gauge
 	taskManager        pooltask.TaskManager[any, any, any, any, pooltask.NilContext]
 	pool.BasePool
@@ -55,6 +56,7 @@ func NewPool(name string, size int32, component util.Component, options ...Optio
 		concurrencyMetrics: metrics.PoolConcurrencyCounter.WithLabelValues(name),
 		taskManager:        pooltask.NewTaskManager[any, any, any, any, pooltask.NilContext](size), // TODO: this general type
 	}
+	result.cond = *sync.NewCond(&result.mu)
 	if size == 0 {
 		return nil, pool.ErrPoolParamsInvalid
 	}
@@ -83,6 +85,7 @@ func (p *Pool) Tune(size int32) {
 // Run runs a function in the pool.
 func (p *Pool) Run(fn func()) error {
 	p.waiting.Add(1)
+	defer p.cond.Signal()
 	defer p.waiting.Add(-1)
 	if p.isStop.Load() {
 		return pool.ErrPoolClosed
@@ -127,6 +130,7 @@ func (p *Pool) run(fn func()) {
 // RunWithConcurrency runs a function in the pool with concurrency.
 func (p *Pool) RunWithConcurrency(fns chan func(), concurrency uint32) error {
 	p.waiting.Add(1)
+	defer p.cond.Signal()
 	defer p.waiting.Add(-1)
 	if p.isStop.Load() {
 		return pool.ErrPoolClosed
@@ -185,7 +189,7 @@ func (p *Pool) ReleaseAndWait() {
 	p.isStop.Store(true)
 	// wait for all the task in the pending to exit
 	for p.waiting.Load() > 0 {
-		_ = "" // it is empty statement to skip empty-block linter.
+		p.cond.Wait()
 	}
 	p.wg.Wait()
 	resourcemanager.InstanceResourceManager.Unregister(p.Name())

--- a/resourcemanager/pool/spool/spool.go
+++ b/resourcemanager/pool/spool/spool.go
@@ -189,9 +189,11 @@ func (p *Pool) checkAndAddRunningInternal(concurrency int32) (conc int32, run bo
 func (p *Pool) ReleaseAndWait() {
 	p.isStop.Store(true)
 	// wait for all the task in the pending to exit
+	p.cond.L.Lock()
 	for p.waiting.Load() > 0 {
 		p.cond.Wait()
 	}
+	p.cond.L.Unlock()
 	p.wg.Wait()
 	resourcemanager.InstanceResourceManager.Unregister(p.Name())
 }

--- a/resourcemanager/pool/spool/spool.go
+++ b/resourcemanager/pool/spool/spool.go
@@ -82,6 +82,8 @@ func (p *Pool) Tune(size int32) {
 
 // Run runs a function in the pool.
 func (p *Pool) Run(fn func()) error {
+	p.waiting.Add(1)
+	defer p.waiting.Add(-1)
 	if p.isStop.Load() {
 		return pool.ErrPoolClosed
 	}
@@ -124,6 +126,8 @@ func (p *Pool) run(fn func()) {
 
 // RunWithConcurrency runs a function in the pool with concurrency.
 func (p *Pool) RunWithConcurrency(fns chan func(), concurrency uint32) error {
+	p.waiting.Add(1)
+	defer p.waiting.Add(-1)
 	if p.isStop.Load() {
 		return pool.ErrPoolClosed
 	}
@@ -145,8 +149,6 @@ func (p *Pool) RunWithConcurrency(fns chan func(), concurrency uint32) error {
 
 // checkAndAddRunning is to check if a task can run. If can, add the running number.
 func (p *Pool) checkAndAddRunning(concurrency uint32) (conc int32, run bool) {
-	p.waiting.Add(1)
-	defer p.waiting.Add(-1)
 	for {
 		if p.isStop.Load() {
 			return 0, false
@@ -183,7 +185,6 @@ func (p *Pool) ReleaseAndWait() {
 	p.isStop.Store(true)
 	// wait for all the task in the pending to exit
 	for p.waiting.Load() > 0 {
-		time.Sleep(waitInterval)
 	}
 	p.wg.Wait()
 	resourcemanager.InstanceResourceManager.Unregister(p.Name())

--- a/resourcemanager/pool/spool/spool.go
+++ b/resourcemanager/pool/spool/spool.go
@@ -42,6 +42,7 @@ type Pool struct {
 	running            atomic.Int32
 	waiting            atomic.Int32
 	isStop             atomic.Bool
+	condMu             sync.Mutex
 	cond               sync.Cond
 	concurrencyMetrics prometheus.Gauge
 	taskManager        pooltask.TaskManager[any, any, any, any, pooltask.NilContext]
@@ -56,7 +57,7 @@ func NewPool(name string, size int32, component util.Component, options ...Optio
 		concurrencyMetrics: metrics.PoolConcurrencyCounter.WithLabelValues(name),
 		taskManager:        pooltask.NewTaskManager[any, any, any, any, pooltask.NilContext](size), // TODO: this general type
 	}
-	result.cond = *sync.NewCond(&result.mu)
+	result.cond = *sync.NewCond(&result.condMu)
 	if size == 0 {
 		return nil, pool.ErrPoolParamsInvalid
 	}

